### PR TITLE
[MM-50283] Allow starting calls in existing thread

### PR DESCRIPTION
**Pull request by streamer45**
#### Summary

PR adds support for starting calls inside threads through the `/call start` slash command. This effectively means that multiple calls could now share the same thread (see screenshot). 

**Note**: to be thorough we should apply similar changes to the mobile side. I'd like to test them a bit on this PR before committing though.

@tanmay-des No UI changes here but pinging you for visibility as this is a somewhat noticeable functionality change that may lead to further discussion.

#### Screenshot

image

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-50283

### CODE REVIEW

File: server/bot_api.go [hunk 1]
```diff
@@ -175,8 +175,8 @@ func (p *Plugin) handleBotPostRecordings(w http.ResponseWriter, r *http.Request,
 		return
 	}
 
-	threadID := info["thread_id"]
-	if threadID == "" {
+	postID := info["thread_id"]
```
Comment (streamer45) [inline server/bot_api.go:178]: This would be a breaking change for recording jobs since they are relying on a proper thread ID to be passed so I decided to handle both cases on this side, although a little less clear but doesn't require us to update everything else.

File: server/plugin.go [hunk 2]
```diff
@@ -318,18 +322,19 @@ func (p *Plugin) startNewCallThread(userID, channelID string, startAt int64, tit
 			return nil, fmt.Errorf("call is missing from channel state")
 		}
 
-		state.Call.ThreadID = createdPost.Id
+		state.Call.PostID = createdPost.Id
+		state.Call.ThreadID = threadID
```
Comment (streamer45) [inline server/plugin.go:326]: Technically this could be optimized since a non-root post would link to its root so we could likely store one field only but I feel it may become more confusing. Happy to hear thoughts though.

Comment (cpoile) [reply server/plugin.go:326]: The only thing that comes to mind is if wrangler or something moves the post?

**Review by cpoile — APPROVED** [review]
Nice!

File: e2e/tests/start_call.spec.ts [hunk 3]
```diff
@@ -77,6 +77,30 @@ test.describe('start new call', () => {
...
+        await post.click();
+        await expect(page.locator('#rhsContainer')).toBeVisible();
+
+        // send slash command in thread to start a call.
+        await page.locator('#reply_textbox').fill('/call start');
+        await page.locator('#reply_textbox').press('Enter');
+        await expect(page.locator('#calls-widget')).toBeVisible();
+
+        // verify the call post is created in the thread.
+        await expect(page.locator('#rhsContainer').filter({has: page.getByText(`${userState.users[0].username} started a call`)})).toBeVisible();
```
[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:98]: I guess it wouldn't hurt, I'll have a look.

File: server/plugin.go (hunk 2 above)
Comment (streamer45) [reply server/plugin.go:326]: Good thought. I guess the edge case would be for that to happen with an on-going call which would likely mess some things. Let me have another pass and see what can be done.

File: e2e/tests/start_call.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:98]: @cpoile Just to clarify, starting a second call from the same thread is allowed. What's not allowed is having two calls at the same time but that goes beyond the changes here, it was never possible to start with. Were you thinking of testing such case?

File: server/plugin.go (hunk 2 above)
Comment (streamer45) [reply server/plugin.go:326]: After thinking a bit about this I am not sure we can do much about this king of edge cases, not easily anyway. Moving a thread means deleting the original so if that happens during a call it would still mess things up since we don't really get notified that the post/thread has been deleted. 

To deal with this we'd probably have to implement a new hook on the server that triggers when a post gets deleted and react accordingly, likely preventing the deletion if a call is on-going since anything else would be even more complex/unexpected (e.g. dropping people from the call).

File: e2e/tests/start_call.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:114]: Added a test for the "starting call twice in same channel" case although not from the call thread since [there's a bug](https://mattermost.atlassian.net/browse/MM-50398) preventing the error from showing.

[the remark this answers is not in the document]
Comment (cpoile) [reply e2e/tests/start_call.spec.ts:114]: Ah, that's weird, I didn't receive a notification for that ping, sorry for the late reply. I meant the former -- just the second call in the same thread. Ohh, https://mattermost.atlassian.net/browse/MM-50398 is a good find. 

Hmm... But why would there be an error in the RHS is we're starting a call from an existing call thread (when the previous call is already finished)?

[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:114]: @cpoile Apologies for the confusion. I am still talking about starting a call with an on-going call, which should error. The other case is allowed as you can have as many subsequent calls as you wish on the same thread since we are essentially creating call posts instead of threads. Do you find that unexpected or is there something I am missing here?

File: server/plugin.go (hunk 2 above)
Comment (cpoile) [reply server/plugin.go:326]: > To deal with this we'd probably have to implement a new hook on the server that triggers when a post gets deleted and react accordingly

Who's working on internalizing wrangler? Cause this seems like an edge-case that we should probably handle, and a "PostWillBeMoved" hook would be really great to add. We could even do that. I'll add a ticket to track it, just in case: https://mattermost.atlassian.net/browse/MM-50408

Comment (streamer45) [reply server/plugin.go:326]: Yes, we need two hooks most likely, a deleted one and a moved one. But it could get tricky with recordings as they could finish after a call has ended so the mapping between previous post id and new post id requires some care.

Also please consider linking or closing https://mattermost.atlassian.net/browse/MM-50397 as I had a similar instinct :)

Comment (cpoile) [reply server/plugin.go:326]: Hah, that's funny. I closed mine and added a comment to yours. :)

File: e2e/tests/start_call.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (cpoile) [reply e2e/tests/start_call.spec.ts:114]: Nope, that's clears it up, and now it seems expected at least to me (but maybe it's needs some PM/design input?).

[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:114]: My thinking was that if we don't allow this behaviour then the whole changes are much less useful. I think one of the originating goals was to keep a recurring thread for subsequent calls but happy to hear thoughts of course as it's easy enough to change.

@itao Let me know if you have concerns/ideas on this.

[the remark this answers is not in the document]
Comment (itao) [reply e2e/tests/start_call.spec.ts:114]: If I understand correctly, the behaviour here is:
- There can still only be one ongoing call per channel at a time
- This lets someone start a call in an existing thread, regardless if the thread was previously used for another call
- If there is already an ongoing call in the channel, regardless if it's in this or another thread, then an error tells the user to join that call instead
- As a user, this lets me escalate from text to voice without losing context of the discussion so far when viewing thread instead of the channel (ie. from threads inbox, RHS, etc)

Is that correct? If so, that makes sense to me and sound great. Could that error for existing call be more helpful by asking if the user would like to join the call instead, as opposed to just flat out error?

[the remark this answers is not in the document]
Comment (streamer45) [reply e2e/tests/start_call.spec.ts:114]: @itao

> - There can still only be one ongoing call per channel at a time
> - This lets someone start a call in an existing thread, regardless if the thread was previously used for another call
> - As a user, this lets me escalate from text to voice without losing context of the discussion so far when viewing thread instead of the channel (ie. from threads inbox, RHS, etc)

Yes to all points.

> - If there is already an ongoing call in the channel, regardless if it's in this or another thread, then an error tells the user to join that call instead

This is where it gets more tricky as we are still relying on a suboptimal errors system native of the webapp. Essentially when a slash command fails for any reason we return an error and then the webapp side handles it by showing it on the bottom of the center channel view:

image

Personally I am not in love with this but historically it was the easiest way to get some sort of feedback to the user so we never tried to do anything better. On top of this there's a [webapp bug](https://mattermost.atlassian.net/browse/MM-50398) preventing such errors from showing if the slash command is sent in a reply box (in a RHS that is). To fix this we'd need changes on the other side.

Now, given the state of things I am starting to think we'd be better implementing our own mechanism for error surfacing, maybe by re-using the existing modals we have for "blocking errors". I created  [MM-50429](https://mattermost.atlassian.net/browse/MM-50429) to review all of this.

**Comment by streamer45** [comment]
E2E failing, why am I not surprised :grimacing:
